src/mqtt_client.rs

```diff
@@ -315,83 +295,117 @@ where
             topic, data, properties
         );
 
+        let ClientContext {
+            session_state,
+            network,
+            ..
+        } = self.sm.context_mut();
+
         // If QoS 0 the ID will be ignored
-        let id = self.session_state.get_packet_identifier();
+        let id = session_state.get_packet_identifier();
 
         let mut buffer: [u8; MSG_SIZE] = [0; MSG_SIZE];
         let packet =
             serialize::publish_message(&mut buffer, topic, data, qos, retain, id, properties)?;
 
-        self.network.write(packet)?;
-        self.session_state.increment_packet_identifier();
+        network.write(packet)?;
+        session_state.increment_packet_identifier();
 
         if qos == QoS::AtLeastOnce {
-            self.session_state.handle_publish(qos, id, packet);
+            session_state.handle_publish(qos, id, packet);
+        }
+
+        Ok(())
```

Comment:
Likewise should move down into `ClientContext`.

Reply:
See above re: subscribe, as we have the same issue here